control-plane/csi-controller/src/controller.rs

```diff
@@ -0,0 +1,726 @@
+                error!("{}", m);
+                Err(Status::internal(m))
+            }
+        }
+        None => {
+            let m = format!("Volume {} reports no current state", volume_id);
+            Err(Status::internal(m))
+        }
+    }
+}
+
+impl From<ApiClientError> for Status {
+    fn from(error: ApiClientError) -> Self {
+        match error {
+            ApiClientError::ResourceNotExists(reason) => Status::not_found(reason),
+            error => Status::internal(format!("Operation failed: {:?}", error)),
+        }
+    }
+}
+
+/// Check whether existing volume is compatible with requested configuration.
+/// Target volume is assumed to exist.
+/// TODO: Add full topology check once Control Plane supports full volume spec.
+#[instrument]
+fn check_existing_volume(
+    volume: &Volume,
+    replica_count: u8,
+    size: u64,
+    pinned_volume: bool,
+) -> Result<(), Status> {
+    // Check if the existing volume is compatible, which means
+    //  - number of replicas is equal or greater
+    //  - size is equal or greater
+    //  - volume is fully created
+    let spec = &volume.spec;
+
+    if spec.status != SpecStatus::Created {
+        return Err(Status::already_exists(format!(
+            "Existing volume {} is in insufficient state: {:?}",
+            spec.uuid, spec.status
+        )));
+    }
+
+    if spec.num_replicas < replica_count {
+        return Err(Status::already_exists(format!(
+            "Existing volume {} has insufficient number of replicas: {} ({} requested)",
+            spec.uuid, spec.num_replicas, replica_count
+        )));
+    }
+
+    if spec.size < size {
+        return Err(Status::already_exists(format!(
+            "Existing volume {} has insufficient size: {} bytes ({} requested)",
+            spec.uuid, spec.size, size
+        )));
+    }
+
+    Ok(())
+}
+
+struct VolumeTopologyMapper {
+    nodes: Vec<Node>,
+}
+
+impl VolumeTopologyMapper {
+    async fn init() -> Result<VolumeTopologyMapper, Status> {
+        let nodes = MayastorApiClient::get_client()
+            .list_nodes()
+            .await
+            .map_err(|e| {
+                Status::failed_precondition(format!(
+                    "Failed to list Mayastor nodes, error = {:?}",
+                    e
+                ))
+            })?;
+
+        Ok(Self { nodes })
+    }
+
+    // Determine the list of nodes where the workload can be placed.
+    // If volume is created as pinned (i.e. local=true), then the nexus and the workload
+    // must be placed on the same node, which in fact means running workloads only on Mayastor
+    // daemonset nodes.
+    // For non-pinned volumes, workload can be put on any node in the Kubernetes cluster.
+    pub fn volume_accessible_topology(&self, pinned_volume: bool) -> Vec<CsiTopology> {
+        if pinned_volume {
+            self.nodes
+                .iter()
+                .map(|n| {
+                    let mut segments = HashMap::new();
```

Comment:
these should probably be only nodes with the openebs.io/engine=mayastor I guess?